vm-wasmi/src/vm.rs

```diff
@@ -505,7 +505,22 @@ where
             .and_then(Extern::into_func)
             .ok_or(WasmiVMError::FunctionNotFound)?;
 
-        export.call(self.0.as_context_mut(), &params, &mut result)?;
+        // Avoid DOS caused by infinite recursion.  See
+        // https://github.com/JumpCrypto/security-research/blob/master/advisories/2023-003-cosmwasm.md?ref=jumpcrypto.com
+        let mut ctx = self.0.as_context_mut();
+        let depth = ctx.data_mut().call_depth_mut();
+        if *depth >= 2 {
+            return Err(WasmiVMError::MaxCallDepthExceeded.into());
+        }
+        *depth += 1;
+
+        let res = export.call(ctx, &params, &mut result);
+
+        let mut ctx = self.0.as_context_mut();
+        let depth = ctx.data_mut().call_depth_mut();
+        *depth = depth.checked_sub(1).unwrap();
```

Comment:
why `checked_sub` if unwrap is used?

Reply:
Perfect. Do you want to ship this `.unwrap()` to prod? Or would you remove it once tests are added?

Reply:
I was planning to keep it.  Don’t have strong feelings either way.

Reply:
So, let's assume that the unwrap gets triggered, what happens @dzmitry-lahoda on the runtime?

Reply:
The runtime crashes but if the unwrap gets triggered than it indicates bug in the code so controlled crash is better than unexpected buggy behaviour later.  Keep in mind this unwrap is unrelated to bug with infinite recursion; this isn’t something malicious contract can trigger.

Reply:
better use exect and ensure you are not panic https://substrate.stackexchange.com/questions/317/when-to-use-expect

Reply:
Honestly I’m unconvinced that `expect` provides any value.  Failing `unwrap` will include file and line which will be enough to identify where the failure happened.  But sure, can change it to `expect`.